### DISCUSSION TITLE
Cant get a session

### DISCUSSION THREAD

**Discussion by Pauleu1** [nextauthjs/next-auth#7390] [Help]
Hi everyone, i have some trouble with my authentication. I created a github repository with the [...nextauth].js. 
When i try to sign up or sign in, the session does not work. I get logged in, but if i refresh the page, or go to a different program on my computer, i get logged out.
Here is my repository, i tried both files. But in the typescript file i also get syntax errors.
https://github.com/Pauleu1/next-auth
I tried a lot of things, but i can't get it to work. I think the session and jwt callbacks are wrong, but i don't know, what i need to change.

**Comment by JasperAlexander** [comment]
You're making it too hard on yourself. Keep it simple and read the docs.

**Reply by Pauleu1** [reply]
So i could use the example and just add my custom claims into the jwt callback or how could i do that? Could you maybe provide some help?

**Comment by OatsProgramming** [comment]
I agree with @JasperAlexander : the code that's in the repo is exceedingly extra (example: setting encode and decode fns in jwt) . However, it is just my opinion and I don't know the full context of your situation. I apologize in advance.

If you want, you can use my code as reference. _Please keep in mind I'm using NextJS (app directory)_

### /api/auth/[...nextauth]/route.ts
```ts
export const authOptions: NextAuthOptions = {
    adapter: PrismaAdapter(prismadb),
    providers: [
        Credentials({
            name: 'Credentials',
            credentials: {
                username: {
                    name: 'username',
                    label: 'Username (optional)',
                    type: 'text',
                    placeholder: 'OatsProgramming'
                },
                email: {
                    name: 'email',
                    label: 'Email (optional)',
                    type: 'email',
                    placeholder: 'jackoatmeals@gmail.com'
                },
                password: {
                    name: 'password',
                    label: 'Password',
                    type: 'password'
                }
            },
            authorize: async (credentials) => {
                if ((!credentials?.username && !credentials?.email) || !credentials?.password) {
                    return null
                }

                try {
                    // Get user from database
                    const user = await prismadb.user.findFirst({
                        where: {
                            OR: [
                                { email: credentials.email },
                                { name: credentials.username }
                            ]
                        }
                    })

                    // See if 404 or password mismatch
                    if (!user) return null
                    else if (!await compare(credentials.password, user.hashedPassword)) return null

                    return {
                        id: user.id,
                        name: user.name,
                        email: user.email,
                        items: user.items,
                    }

                } catch (err) {
                    console.error(err)
                    return null
                }
            }
        })
    ],
    callbacks: {
        session: ({ session, token }) => {
            console.log(session)
            return {
                ...session,
                user: {
                    ...session.user,
                    id: token.id,
                    items: token.items
                }
            }
        },
        jwt: ({ user, token }) => {
            console.log(token)
            if (user) {
                return {
                    ...token,
                    ...user,
                }
            }
            return token
        }
    },
    session: {
        strategy: 'jwt'
    }
}

const handler = NextAuth(authOptions)
export { handler as GET, handler as POST }
```

### Any Server Component That You Want to Have Session
 ```ts
export default async function ServerComponent() {
  const session = await getServerSession(authOptions)
  .....
}

**Comment by OatsProgramming** [comment]
_Provided answer got too long that I have to separate it_

### If Client Side
If you're planning to use the session on the client side, you'll have to use a SessionProvider. It's basically useContext from React but for session.

_I recommend putting this near the root_

```ts
'use client'

import { SessionProvider } from "next-auth/react"

export default function AuthProvider({ children }: {
    children: React.ReactNode
}) {
    return (
        <SessionProvider>
            {children}
        </SessionProvider>
    )
}
```

Then you can use it anywhere in the app
```ts
'use client'

import { useSession } from "next-auth/react";

export default function Page() {
    const { data: session } = useSession()

    return (
        <>
            <h1>Client Session</h1>
            <pre>{JSON.stringify(session)}</pre>
        </>
    )
}
```

**Comment by Pauleu1** [comment]
Hi Jack, i managed to fix my issue today. But i have a different problem now. Is it normal, that the jwt callback is executed first, then the session callback. After that the jwt callback is executed again. The problem is, that the id, which i need for my graphql query is included only if the session callback is executed a second time.

**Comment by OatsProgramming** [comment]
@Pauleu1 Well, to reiterate from the docs, yes: jwt() is executed first then session() when you're using jwt. Here's the docs for reference to that [https://next-auth.js.org/configuration/callbacks#session-callback](https://next-auth.js.org/configuration/callbacks#session-callback).
_(It's highlighted in green)_

As for the id issue, it's something that I need to see the code for. Can't tell if it's a bug within the code or bad code.
